Seafile WebDAV extension (dependencies, package, and integration to seafile service)

flyx opened the pull request:
###### Description of changes

This adds the Python library `seafobj` and the Python WSGI application `seafdav`, both of which belong with the [Seafile](https://manual.seafile.com/extension/webdav/) server. Version is `8.0.8` which is not the most recent one, but conforms to the current versions of existing `seafile-server` and `seahub` packages.

`seafdav` provides a WebDAV interface to a running `seafile-server`. `seafobj` is a dependency of `seafdav`.

The `nixos/seafile` service has been extended to provide options for automatically starting the WebDAV extension alongside the server. I opted to not allow freeform settings since the existing settings are already confusing – some are consumed by `seafdav` itself, some by the script `server_cli` (which my setup doesn't use), and some only when using Seafile's own management script `seafile.sh` (which isn't viable for NixOS). I wanted to clearly communicate which values are actually relevant.

I reviewed the relevant code (mainly [server_cli.py](https://github.com/haiwen/seafdav/blob/master/wsgidav/server/server_cli.py)) to learn how the settings are used, and then added and used a simpler WSGI application setup script for the service setup. This enables a service setup very similar to the one of `seahub`, which also is a WSGI application. This should be easier to maintain than running the `server_cli.py` script provided from upstream.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

ping @schmittlauch @greizgh (maintainers of the other Seafile packages)

flyx committed:
maintainers: add flyx

flyx committed:
seafobj: init at 9.0.7

Python module required for the Seafile WebDAV extension

flyx committed:
seafdav: init at 9.0.7

WebDAV extension for Seafile server.
Package adds a WSGI application for easy deployment, which is
lacking from upstream sources.

flyx committed:
nixos/seafile: support running WebDAV extension

Added option to start the optional WebDAV service alongside a
Seafile server. Added options for configuring the WebDAV service.
Added subtests for WebDAV server to Seafile test.

pbsds commented on pkgs/applications/networking/seafdav/default.nix:
```diff
@@ -0,0 +1,54 @@
+  src = fetchFromGitHub {
+    owner = "haiwen";
+    repo = "seafdav";
+    rev = "v8.0.8-server";
```
```suggestion
    rev = "v${version}-server";
```

pbsds reviewed (changes requested):
Impressive PR for a first-time contributor! Sorry for the long wait, this is not normal. Do not be discouraged!

The diff LGTM, but i have not test-run the package.

Could you please consider adding a nixos test for this module? Something as simple as a test that pings the webdav port should be sufficient.
[here is an example](https://github.com/NixOS/nixpkgs/blob/8473365f698a94cf1729b3273e8ef778d0607a29/nixos/tests/polaris.nix#L26-L28)

pbsds commented:
Result of `nixpkgs-review pr 171878` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>

flyx commented:
@pbsds Thanks for the review. I addressed your comments. Since Seafile has already a test, I simply added WebDAV subtests to it. Testing those showed my I forgot adding my packages to `top-level`, which I fixed. Running the test runs into an error in the seahub part (missing module 'drafts') and I haven't been able to figure out why, but perhaps I am simply running the test the wrong way…

greizgh commented on nixos/modules/services/networking/seafile.nix:
```diff
@@ -282,6 +330,42 @@ in {
+        requires = [ "seaf-server.service" ];
+        restartTriggers = [ seahubSettings ];
+        environment = {
+          PYTHONPATH = "${pkgs.seafdav.pythonPath}:${pkgs.seafdav}:/etc/seafile";
```
Do we need `/etc/seafile` in pythonpath?
```suggestion
          PYTHONPATH = "${pkgs.seafdav.pythonPath}:${pkgs.seafdav}";
```

flyx replied:
This is required for loading the `seahub_settings.py` which is processed by SeafDAV. From what I understand, this is handled via django settings for actual seahub, but that is not used in SeafDAV so I'm unsure whether there's a better way to do this.

greizgh commented on pkgs/applications/networking/seafdav/default.nix:
```diff
@@ -0,0 +1,54 @@
+  src = fetchFromGitHub {
+    owner = "haiwen";
+    repo = "seafdav";
+    rev = "v${version}-server";
```
Upstream has a habit of pushing over existing tags, on other seafile component we end up using commit hashes.
Let's hope this is not the case here :crossed_fingers:

greizgh commented on pkgs/applications/networking/seafdav/default.nix:
```diff
@@ -0,0 +1,54 @@
+}:
+python3.pkgs.buildPythonApplication rec {
+  pname = "seafdav";
+  version = "8.0.8";
```
seafile-server has been updated to 9.0.x, maybe update this component too?

flyx replied:
Yes now that the other packages have been updated this should be too. I pushed the `9.0.7` version.

greizgh reviewed (commented):
re: testing, draft error sounds like #173825. Make sure you rebase to get the fix.

flyx commented:
@greizgh Did a rebase and the error vanished, but test still fails with a timeout on starting seahub. Might be that my machine is simply too slow.

greizgh commented:
> @greizgh Did a rebase and the error vanished, but test still fails with a timeout on starting seahub. Might be that my machine is simply too slow.

Tests were flaky because of sleep conditions, it should be fixed by https://github.com/NixOS/nixpkgs/pull/195253.

ddelabru commented:
This addresses the same functionality as #152069 but at present it looks like this one is the more complete pull request.

hustlerone commented:
So... What's keeping this from being merged?

flyx commented:
> So... What's keeping this from being merged?

I lost interest in pushing this to get merged due to the long process. I just integrated it as local overlay on my system.

For the PR to be merged the seafobj/seafdav version should be updated (server and seahub have been bumped to `11.0.12` on nixos stable & unstable). Not sure whether this requires additional changes. In any case, I won't be doing it since the newer Seafile versions dropped support for postgres, which make me rather want to switch to an alternative than to upgrade.

Feel free to adopt the PR.

hustlerone commented:
Nextcloud isn't any better hence my interest

mdaniels5757 commented:
The `seafile` module and the `seahub` and `seafile-server` packages were dropped in #435059. Not sure what that means for this PR.
